gateway/src/project.rs

```diff
@@ -0,0 +1,613 @@
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Project {
+    Creating(ProjectCreating),
+    Starting(ProjectStarting),
+    Started(ProjectStarted),
+    Stopped(ProjectStopped),
+    Errored(ProjectError),
+}
+
+impl_from_variant!(Project:
+                   ProjectCreating => Creating,
+                   ProjectStarting => Starting,
+                   ProjectStarted => Started,
+                   ProjectStopped => Stopped,
+                   ProjectError => Errored);
+
+impl Project {
+    pub async fn stop<'c, C: Context<'c>>(self, ctx: &C) -> Result<Self, Error> {
+        match self {
+            Self::Creating(_) => Err(Error::custom(
+                ErrorKind::InvalidOperation,
+                "tried to stop a project that was not ready",
+            )),
+            Self::Starting(ProjectStarting { container, .. }) => {
+                Ok(Self::Stopped(ProjectStopped { container }))
+            }
+            Self::Started(ProjectStarted { container, .. }) => {
+                ctx.docker()
+                    .stop_container(container.id.as_ref().unwrap(), None)
+                    .await?;
+                Ok(Self::Stopped(ProjectStopped {
+                    container: container.refresh(ctx).await?,
+                }))
+            }
+            Self::Stopped(stopped) => Ok(Self::Stopped(stopped)),
+            Self::Errored(err) => Ok(Self::Errored(err)),
+        }
+    }
+
+    pub fn target_ip(&self) -> Result<Option<String>, Error> {
+        match self.clone() {
+            Self::Started(project_started) => Ok(Some(project_started.target_ip().to_string())),
+            _ => Ok(None), // not ready
+        }
+    }
+
+    pub fn state(&self) -> &'static str {
+        match self {
+            Self::Started(_) => "started",
+            Self::Stopped(_) => "stopped",
+            Self::Starting(_) => "starting",
+            Self::Creating(_) => "creating",
+            Self::Errored(_) => "error",
+        }
+    }
+
+    pub fn container_id(&self) -> Option<String> {
+        match self {
+            Self::Starting(ProjectStarting {
+                container: ContainerInspectResponse { id, .. },
+                ..
+            })
+            | Self::Started(ProjectStarted {
+                container: ContainerInspectResponse { id, .. },
+                ..
+            })
+            | Self::Stopped(ProjectStopped {
+                container: ContainerInspectResponse { id, .. },
+            }) => id.clone(),
+            Self::Errored(ProjectError { ctx: Some(ctx), .. }) => ctx.container_id(),
+            _ => None,
+        }
+    }
+
+    pub async fn destroy<'c, C: Context<'c>>(self, ctx: &C) -> Result<(), Error> {
+        if let Some(ref container_id) = self.container_id() {
+            ctx.docker()
+                .stop_container(container_id, Some(StopContainerOptions { t: 30 }))
+                .await
+                .unwrap_or(());
```

Comment:
A matter of preference but `let _ = ` appears to be the more common way to discard a `Result` at least in my experience.
```suggestion
            let _ = ctx.docker()
                .stop_container(container_id, Some(StopContainerOptions { t: 30 }))
                .await;
```